gmail/src/services/odoo_auth.ts

```diff
@@ -0,0 +1,133 @@
+ * 2. The user is redirected to Odoo and enter his login / password
+ * 3. Then the user is redirected to the Google App-Script
+ * 4. Thanks the the state token, the function "odooAuthCallback" is called with the auth code
+ * 5. The auth code is exchanged for an access token with a RPC call
+ */
+function odooAuthCallback(callbackRequest: any) {
+    Logger.log("Run authcallback");
+    const success = callbackRequest.parameter.success;
+    const authCode = callbackRequest.parameter.auth_code;
+
+    if (success !== "1") {
+        return HtmlService.createHtmlOutput(
+            errorPage.replace("__ERROR_MESSAGE__", "Odoo did not return successfully."),
+        );
+    }
+
+    Logger.log("Get access token from auth code...");
+
+    const userProperties = PropertiesService.getUserProperties();
+    const odooUrl = userProperties.getProperty("ODOO_SERVER_URL");
+
+    const response = postJsonRpc(odooUrl + ODOO_AUTH_URLS.codeValidation, {
+        auth_code: authCode,
+    });
+
+    if (!response || !response.access_token || !response.access_token.length) {
+        return HtmlService.createHtmlOutput(
+            errorPage.replace(
+                "__ERROR_MESSAGE__",
+                "The token exchange failed. Maybe your token has expired or your database can not be reached by the Google server." +
+                    "<hr noshade>Contact your administrator or our support.",
+            ),
+        );
+    }
+
+    const accessToken = response.access_token;
+
+    userProperties.setProperty("ODOO_ACCESS_TOKEN", accessToken);
+
+    return HtmlService.createHtmlOutput("Success ! <script>top.window.close()</script>");
+}
+
+/**
+ * Generate the URL to redirect the user for the authentication to the Odoo database.
+ *
+ * This URL contains a state and the Odoo database should resend it.
+ * The Google server use the state code to now which function to execute when the user
+ * is redirected on their server.
+ */
+export function getServiceUrl() {
+    const userProperties = PropertiesService.getUserProperties();
+    const odooUrl = userProperties.getProperty("ODOO_SERVER_URL");
+    const scriptId = ScriptApp.getScriptId();
+
+    if (!odooUrl || !odooUrl.length) {
+        throw new Error("Can not retrieve the Odoo database URL.");
+    }
+
+    if (!scriptId || !scriptId.length) {
+        throw new Error("Can not retrieve the script ID.");
+    }
+
+    const stateToken = ScriptApp.newStateToken().withMethod("odooAuthCallback").withTimeout(3600).createToken();
+
+    const redirectToAddon = `https://script.google.com/macros/d/${scriptId}/usercallback`;
+    const scope = ODOO_AUTH_URLS.scope;
+
+    const url =
+        odooUrl +
+        ODOO_AUTH_URLS.authCodePage +
+        "?redirect=" +
+        encodeURIComponent(redirectToAddon) +
+        "&friendlyname=Gmail" +
+        "&state=" +
+        encodeURIComponent(stateToken) +
+        "&scope=" +
+        encodeURIComponent(scope);
```

Comment:
Could use template string here as well.
Isn't there a nice JS way to add parameters to an URL using a JSON object?

Reply:
I've done my own method to make the code cleaner :)